File: apps/server/internal/sanitize/adversarial_test.go
package sanitize_test

import (
	"fmt"
	"math/rand"
	"net/url"
	"strings"
	"testing"

	"github.com/dannyota/aboutme/apps/server/internal/sanitize"
	schemagen "github.com/dannyota/aboutme/packages/schema/gen/go"
	"golang.org/x/net/html"
	"golang.org/x/net/html/atom"
)

const (
	blindPropertySeed  int64 = 0x5033474f5a17
	blindPropertyCases       = 2048
)

type blindPolicy struct {
	allowedTags       map[string]struct{}
	allowedAttributes map[string]map[string]struct{}
	allowedSchemes    map[string]struct{}
	forbiddenPrefixes []string
	externalRel       string
}

func newBlindPolicy(t *testing.T) blindPolicy {
	t.Helper()

	policy := blindPolicy{
		allowedTags:       make(map[string]struct{}, len(schemagen.AllowedTags)),
		allowedAttributes: make(map[string]map[string]struct{}, len(schemagen.AllowedAttributes)),
		allowedSchemes:    make(map[string]struct{}, len(schemagen.AllowedURLSchemes)),
		forbiddenPrefixes: make([]string, 0, len(schemagen.ForbiddenAttributePrefixes)),
		externalRel:       schemagen.ExternalRel,
	}
	for _, tag := range schemagen.AllowedTags {
		tag = strings.ToLower(tag)
		if _, duplicate := policy.allowedTags[tag]; duplicate {
			t.Fatalf("blind policy: duplicate allowed tag %q", tag)
		}
		policy.allowedTags[tag] = struct{}{}
	}
	for tag, attributes := range schemagen.AllowedAttributes {
		tag = strings.ToLower(tag)
		allowed := make(map[string]struct{}, len(attributes))
		for _, attribute := range attributes {
			attribute = strings.ToLower(attribute)
			if _, duplicate := allowed[attribute]; duplicate {
				t.Fatalf("blind policy: duplicate %q attribute %q", tag, attribute)
			}
			allowed[attribute] = struct{}{}
		}
		policy.allowedAttributes[tag] = allowed
	}
	for _, scheme := range schemagen.AllowedURLSchemes {
		scheme = strings.ToLower(scheme)
		if _, duplicate := policy.allowedSchemes[scheme]; duplicate {
			t.Fatalf("blind policy: duplicate allowed URL scheme %q", scheme)
		}
		policy.allowedSchemes[scheme] = struct{}{}
	}
	for _, prefix := range schemagen.ForbiddenAttributePrefixes {
		policy.forbiddenPrefixes = append(policy.forbiddenPrefixes, strings.ToLower(prefix))
	}

	return policy
}

func (p blindPolicy) check(fragment string) error {
	nodes, err := parseFragment(fragment)
	if err != nil {
		return fmt.Errorf("parse fragment: %w", err)
	}
	for _, node := range nodes {
		if err := p.checkNode(node); err != nil {
			return err
		}
	}
	return nil
}

func (p blindPolicy) checkNode(node *html.Node) error {
	switch node.Type {
	case html.TextNode, html.CommentNode:
		return nil
	case html.ElementNode:
		if node.Namespace != "" {
			return fmt.Errorf("foreign namespace %q on <%s>", node.Namespace, node.Data)
		}
		tag := strings.ToLower(node.Data)
		if _, ok := p.allowedTags[tag]; !ok {
			return fmt.Errorf("tag <%s> is not allowed", node.Data)
		}

		values := make(map[string]string, len(node.Attr))
		for _, attribute := range node.Attr {
			if attribute.Namespace != "" {
				return fmt.Errorf("attribute namespace %q on <%s %s>", attribute.Namespace, tag, attribute.Key)
			}
			key := strings.ToLower(attribute.Key)
			if _, duplicate := values[key]; duplicate {
				return fmt.Errorf("duplicate attribute %q on <%s>", key, tag)
			}
			for _, prefix := range p.forbiddenPrefixes {
				if strings.HasPrefix(key, prefix) {
					return fmt.Errorf("attribute %q on <%s> has forbidden prefix %q", key, tag, prefix)
				}
			}
			if _, ok := p.allowedAttributes[tag][key]; !ok {
				return fmt.Errorf("attribute %q is not allowed on <%s>", key, tag)
			}
			values[key] = attribute.Val
		}

		if tag == "a" {
			rel, ok := values["rel"]
			if !ok {
				return fmt.Errorf("<a> is missing rel")
			}
			if rel != p.externalRel {
				return fmt.Errorf("<a> rel is %q, want %q", rel, p.externalRel)
			}
			if target, ok := values["target"]; ok && target != "_blank" {
				return fmt.Errorf("<a> target is %q, want _blank or absent", target)
			}
			if href, ok := values["href"]; ok {
				if err := p.checkURL(href); err != nil {
					return fmt.Errorf("<a> href %q: %w", href, err)
				}
			}
		}

		for child := node.FirstChild; child != nil; child = child.NextSibling {
			if err := p.checkNode(child); err != nil {
				return err
			}
		}
		return nil
	default:
		return fmt.Errorf("node type %d is not admitted", node.Type)
	}
}

func (p blindPolicy) checkURL(raw string) error {
	// The URL Standard removes leading/trailing C0 controls and spaces, and
	// removes ASCII tab/newline/carriage-return characters before parsing.
	// Apply that browser-facing normalization before deciding the scheme.
	normalized := normalizeBrowserURL(raw)
	parsed, err := url.Parse(normalized)
	if err != nil {
		return fmt.Errorf("URL does not parse: %w", err)
	}
	scheme := strings.ToLower(parsed.Scheme)
	if scheme == "" {
		return fmt.Errorf("URL has no explicit scheme")
	}
	if _, ok := p.allowedSchemes[scheme]; !ok {
		return fmt.Errorf("scheme %q is not allowed", scheme)
	}
	return nil
}

func normalizeBrowserURL(raw string) string {
	start, end := 0, len(raw)
	for start < end && raw[start] <= 0x20 {
		start++
	}
	for end > start && raw[end-1] <= 0x20 {
		end--
	}

	var normalized strings.Builder
	normalized.Grow(end - start)
	for i := start; i < end; i++ {
		switch raw[i] {
		case '\t', '\n', '\r':
			continue
		default:
			normalized.WriteByte(raw[i])
		}
	}
	return normalized.String()
}

func parseFragment(fragment string) ([]*html.Node, error) {
	context := &html.Node{Type: html.ElementNode, DataAtom: atom.Div, Data: "div"}
	return html.ParseFragment(strings.NewReader(fragment), context)
}

func parserRoundTrip(fragment string) (string, error) {
	nodes, err := parseFragment(fragment)
	if err != nil {
		return "", err
	}

	var rendered strings.Builder
	for _, node := range nodes {
		if err := html.Render(&rendered, node); err != nil {
			return "", err
		}
	}
	return rendered.String(), nil
}

func TestBlindNeutralizationPredicateNegativeControls(t *testing.T) {
	policy := newBlindPolicy(t)

	unsafeCorpusCases := 0
	for _, payload := range schemagen.HostileCorpus {
		err := policy.check(payload.Payload)
		if payload.ID == "js-scheme-bare" {
			if err != nil {
				t.Errorf("raw corpus %q is dangerous-looking text, but the predicate rejected it: %v", payload.ID, err)
			}
			continue
		}
		unsafeCorpusCases++
		if err == nil {
			t.Errorf("raw corpus %q contains an active policy violation, but the predicate accepted it", payload.ID)
		}
	}
	if unsafeCorpusCases == 0 {
		t.Fatal("negative control exercised no unsafe corpus cases")
	}

	violations := map[string]string{
		"forbidden element":     `<script>alert(1)</script>`,
		"forbidden prefix":      `<p onclick="alert(1)">text</p>`,
		"unlisted attribute":    `<p style="background:url(javascript:alert(1))">text</p>`,
		"forbidden scheme":      `<a href="javascript:alert(1)" rel="noopener noreferrer">text</a>`,
		"relative URL":          `<a href="/relative" rel="noopener noreferrer">text</a>`,
		"protocol-relative URL": `<a href="//example.test" rel="noopener noreferrer">text</a>`,
		"missing rel":           `<a href="https://example.test">text</a>`,
		"forged rel":            `<a href="https://example.test" rel="opener">text</a>`,
		"forged target":         `<a href="https://example.test" rel="noopener noreferrer" target="_self">text</a>`,
		"foreign namespace":     `<svg><a href="https://example.test">text</a></svg>`,
	}
	for name, fragment := range violations {
		t.Run(name, func(t *testing.T) {
			if err := policy.check(fragment); err == nil {
				t.Fatalf("predicate accepted hand-built violation %q", fragment)
			}
		})
	}

	bareText := `javascript:alert(1) &lt;img src=x onerror=alert(1)&gt; data:text/html,&lt;script&gt;alert(1)&lt;/script&gt;`
	if err := policy.check(bareText); err != nil {
		t.Fatalf("predicate rejected dangerous-looking bare text: %v", err)
	}
	nodes, err := parseFragment(bareText)
	if err != nil {
		t.Fatalf("parse dangerous-looking bare text: %v", err)
	}
	if len(nodes) != 1 || nodes[0].Type != html.TextNode {
		t.Fatalf("dangerous-looking bare text parsed as active markup: %#v", nodes)
	}
}

func TestRichTextBlindHostileCorpus(t *testing.T) {
	policy := newBlindPolicy(t)
	if sanitize.AllowlistVersion != schemagen.SanitizerAllowlistVersion {
		t.Fatalf("sanitizer policy version = %d, generated policy version = %d", sanitize.AllowlistVersion, schemagen.SanitizerAllowlistVersion)
	}

	for _, payload := range schemagen.HostileCorpus {
		t.Run(payload.ID, func(t *testing.T) {
			assertSanitizerProperties(t, policy, payload.Payload)
		})
	}
}

func TestRichTextBlindSpecDerivedPayloads(t *testing.T) {
	policy := newBlindPolicy(t)
	payloads := map[string]string{
		"mXSS noscript pivot":          `<noscript><p title="</noscript><img src=x onerror=alert(1)>">pivot</p></noscript>`,
		"mXSS template pivot":          `<template><p><svg><foreignObject><style><!--</style><img title="--><img src=x onerror=alert(1)>"></foreignObject></svg></p></template>`,
		"mXSS foreign-content pivot":   `<math><mtext><table><mglyph><style><!--</style><img title="--><img src=x onerror=alert(1)>"></mglyph></table></mtext></math>`,
		"URL-encoded colon":            `<a href="javascript%3Aalert(1)" rel="noopener noreferrer">encoded</a>`,
		"mixed entity and case":        `<a href="JaVa&#x53;CrIpT&#x3a;alert(1)" rel="noopener noreferrer">mixed</a>`,
		"entity and whitespace scheme": `<a href="&#x09;DaTa&#58;text/html,&lt;script&gt;alert(1)&lt;/script&gt;" rel="noopener noreferrer">mixed</a>`,
		"formaction smuggling":         `<form><button formaction="javascript:alert(1)">submit</button></form>`,
		"srcdoc smuggling":             `<iframe srcdoc="&lt;script&gt;alert(1)&lt;/script&gt;"></iframe>`,
		"xlink href smuggling":         `<svg><a xlink:href="javascript:alert(1)"><text>svg link</text></a></svg>`,
		"style smuggling":              `<p style="background-image:url(javascript:alert(1))">styled</p>`,
		"math namespace confusion":     `<math><mtext><a href="javascript:alert(1)">math link</a></mtext></math>`,
		"SVG namespace confusion":      `<svg><a href="https://example.test" rel="noopener noreferrer"><text>svg link</text></a></svg>`,
		"rel token forgery":            `<a href="https://example.test" rel="noopener noreferrer opener" target="_blank">forged</a>`,
		"target forgery":               `<a href="https://example.test" rel="noopener noreferrer" target="_parent">forged</a>`,
		"attribute duplication":        `<a href="javascript:alert(1)" href="https://example.test" rel="noopener noreferrer">duplicate</a>`,
	}

	for name, payload := range payloads {
		t.Run(name, func(t *testing.T) {
			if err := policy.check(payload); err == nil {
				t.Fatalf("spec-derived adversarial input unexpectedly satisfies the blind predicate: %q", payload)
			}
			assertSanitizerProperties(t, policy, payload)
		})
	}
}

func TestRichTextBlindDeterministicProperties(t *testing.T) {
	policy := newBlindPolicy(t)
	random := rand.New(rand.NewSource(blindPropertySeed))

	for caseIndex := 0; caseIndex < blindPropertyCases; caseIndex++ {
		input := randomHTMLLikeString(random)
		output := sanitize.RichText(input)
		if err := policy.check(output); err != nil {
			t.Fatalf("seed=%#x case=%d sanitizer output violates policy: %v\ninput=%q\noutput=%q", blindPropertySeed, caseIndex, err, input, output)
		}
		if second := sanitize.RichText(output); second != output {
			t.Fatalf("seed=%#x case=%d sanitizer is not idempotent\ninput=%q\nfirst=%q\nsecond=%q", blindPropertySeed, caseIndex, input, output, second)
		}

		parserOutput, err := parserRoundTrip(output)
		if err != nil {
			t.Fatalf("seed=%#x case=%d parser round trip failed: %v\noutput=%q", blindPropertySeed, caseIndex, err, output)
		}
		if err := policy.check(parserOutput); err != nil {
			t.Fatalf("seed=%#x case=%d parser round trip reintroduced a violation: %v\ninput=%q\noutput=%q\nround_trip=%q", blindPropertySeed, caseIndex, err, input, output, parserOutput)
		}
		crossFeed := sanitize.RichText(parserOutput)
		if err := policy.check(crossFeed); err != nil {
			t.Fatalf("seed=%#x case=%d parser output fed back through the public sanitizer violates policy: %v\nround_trip=%q\ncross_feed=%q", blindPropertySeed, caseIndex, err, parserOutput, crossFeed)
		}
		if second := sanitize.RichText(crossFeed); second != crossFeed {
			t.Fatalf("seed=%#x case=%d parser-fed output is not idempotent\nfirst=%q\nsecond=%q", blindPropertySeed, caseIndex, crossFeed, second)
		}
	}
}

func assertSanitizerProperties(t *testing.T, policy blindPolicy, input string) {
	t.Helper()

	output := sanitize.RichText(input)
	if err := policy.check(output); err != nil {
		t.Fatalf("sanitizer output violates blind policy: %v\ninput=%q\noutput=%q", err, input, output)
	}
	if second := sanitize.RichText(output); second != output {
		t.Fatalf("sanitizer is not idempotent\ninput=%q\nfirst=%q\nsecond=%q", input, output, second)
	}

	parserOutput, err := parserRoundTrip(output)
	if err != nil {
		t.Fatalf("parser round trip: %v", err)
	}
	if err := policy.check(parserOutput); err != nil {
		t.Fatalf("parser round trip reintroduced a violation: %v\ninput=%q\noutput=%q\nround_trip=%q", err, input, output, parserOutput)
	}
	crossFeed := sanitize.RichText(parserOutput)
	if err := policy.check(crossFeed); err != nil {
		t.Fatalf("parser output fed back through the public sanitizer violates policy: %v\nround_trip=%q\ncross_feed=%q", err, parserOutput, crossFeed)
	}
}

func randomHTMLLikeString(random *rand.Rand) string {
	fragments := [...]string{
		`<p>`, `</p>`, `<br>`, `<strong>`, `</strong>`, `<em>`, `</em>`,
		`<a href="https://example.test" rel="noopener noreferrer" target="_blank">`, `</a>`,
		`<script>`, `</script>`, `<style>`, `</style>`, `<template>`, `</template>`,
		`<noscript>`, `</noscript>`, `<svg>`, `</svg>`, `<math>`, `</math>`,
		` href="javascript:alert(1)"`, ` href="//example.test"`, ` href="javascript%3Aalert(1)"`,
		` onclick="alert(1)"`, ` style="background:url(javascript:alert(1))"`,
		` srcdoc="&lt;script&gt;alert(1)&lt;/script&gt;"`, ` xlink:href="javascript:alert(1)"`,
		` rel="opener"`, ` target="_self"`, `<!--`, `-->`, `&lt;`, `&#x3a;`, `&amp;`,
		"\x00", "\t", "\n", "\r", "javascript:alert(1)", "data:text/html,",
	}

	parts := random.Intn(48)
	var generated strings.Builder
	for i := 0; i < parts; i++ {
		switch random.Intn(4) {
		case 0, 1:
			generated.WriteString(fragments[random.Intn(len(fragments))])
		case 2:
			generated.WriteByte(byte(random.Intn(256)))
		case 3:
			length := 1 + random.Intn(12)
			for j := 0; j < length; j++ {
				generated.WriteByte(byte(random.Intn(256)))
			}
		}
	}
	return generated.String()
}

File: apps/web/test/sanitizer/adversarial.test.ts
// @vitest-environment jsdom

import {
  ALLOWED_ATTRIBUTES,
  ALLOWED_TAGS,
  ALLOWED_URL_SCHEMES,
  EXTERNAL_REL,
  FORBIDDEN_ATTRIBUTE_PREFIXES,
  HOSTILE_CORPUS,
} from '@aboutme/schema/sanitizer';
import { describe, expect, it } from 'vitest';
import { sanitizeRichText } from '../../app/utils/sanitizeRichText';

const HTML_NAMESPACE = 'http://www.w3.org/1999/xhtml';
const ARBITRARY_STRING_SEED = 0x51a17e3d;
const ARBITRARY_STRING_CASES = 512;

const allowedTags = new Set<string>(ALLOWED_TAGS);
const allowedSchemes = new Set<string>(ALLOWED_URL_SCHEMES);
const expectedRelTokens = EXTERNAL_REL.split(/\s+/);

type Rule = 'attr' | 'href' | 'ns' | 'rel' | 'tag' | 'target';

interface NeutralizationViolation {
  readonly detail: string;
  readonly rule: Rule;
}

function parseRichText(html: string): Document {
  return new DOMParser().parseFromString(html, 'text/html');
}

function trimC0ControlsAndSpace(value: string): string {
  let start = 0;
  let end = value.length;

  while (start < end && value.charCodeAt(start) <= 0x20) {
    start += 1;
  }
  while (end > start && value.charCodeAt(end - 1) <= 0x20) {
    end -= 1;
  }

  return value.slice(start, end);
}

function hasAllowedExplicitScheme(href: string): boolean {
  // Match the URL parser's removal of tabs/newlines and outer C0 controls
  // before checking that the source supplied an explicit admitted scheme.
  const normalized = trimC0ControlsAndSpace(href.replace(/[\t\n\r]/g, ''));
  const match = /^([a-z][a-z\d+.-]*):/i.exec(normalized);

  return match !== null && allowedSchemes.has(match[1].toLowerCase());
}

function hasExactExternalRel(rel: string | null): boolean {
  if (rel === null) {
    return false;
  }

  const actual = rel
    .trim()
    .split(/\s+/)
    .filter(Boolean)
    .map((token) => token.toLowerCase());

  const hasExpectedCount = actual.length === expectedRelTokens.length;
  const hasDistinctTokens = new Set(actual).size === expectedRelTokens.length;
  const hasExpectedTokens = expectedRelTokens.every((token) =>
    actual.includes(token),
  );

  return hasExpectedCount && hasDistinctTokens && hasExpectedTokens;
}

function neutralizationViolations(html: string): NeutralizationViolation[] {
  const document = parseRichText(html);
  const violations: NeutralizationViolation[] = [];

  for (const scaffold of [
    document.documentElement,
    document.head,
    document.body,
  ]) {
    for (const attribute of Array.from(scaffold.attributes)) {
      violations.push({
        detail: `${scaffold.localName} retained ${attribute.name}`,
        rule: 'attr',
      });
    }
  }

  function visit(node: Node): void {
    if (node.nodeType === Node.ELEMENT_NODE) {
      const element = node as Element;
      const tag = element.localName.toLowerCase();

      if (element.namespaceURI !== HTML_NAMESPACE) {
        violations.push({
          detail: `${tag} namespace ${element.namespaceURI ?? 'null'}`,
          rule: 'ns',
        });
      }
      if (!allowedTags.has(tag)) {
        violations.push({
          detail: `<${tag}> is outside the allowlist`,
          rule: 'tag',
        });
      }

      const tagAttributes = new Set<string>(ALLOWED_ATTRIBUTES[tag] ?? []);
      for (const attribute of Array.from(element.attributes)) {
        const name = attribute.name.toLowerCase();
        const hasForbiddenPrefix = FORBIDDEN_ATTRIBUTE_PREFIXES.some((prefix) =>
          name.startsWith(prefix.toLowerCase()),
        );

        if (hasForbiddenPrefix || !tagAttributes.has(name)) {
          violations.push({
            detail: `<${tag}> retained disallowed attribute ${attribute.name}`,
            rule: 'attr',
          });
        }
      }

      if (tag === 'a') {
        const href = element.getAttribute('href');
        if (href !== null && !hasAllowedExplicitScheme(href)) {
          violations.push({
            detail: `<a> retained disallowed href: ${href}`,
            rule: 'href',
          });
        }
        if (!hasExactExternalRel(element.getAttribute('rel'))) {
          violations.push({
            detail: `<a> did not receive the exact hardened rel token set`,
            rule: 'rel',
          });
        }

        const target = element.getAttribute('target');
        if (target !== null && target !== '_blank') {
          violations.push({
            detail: `<a> retained non-_blank target ${target}`,
            rule: 'target',
          });
        }
      }

      if (tag === 'template' && element.namespaceURI === HTML_NAMESPACE) {
        visit((element as HTMLTemplateElement).content);
      }
    }

    for (const child of Array.from(node.childNodes)) {
      visit(child);
    }
  }

  for (const root of [
    ...document.head.childNodes,
    ...document.body.childNodes,
  ]) {
    visit(root);
  }

  return violations;
}

function isNeutralized(html: string): boolean {
  return neutralizationViolations(html).length === 0;
}

function anchorsIn(html: string): HTMLAnchorElement[] {
  return Array.from(parseRichText(html).querySelectorAll('a'));
}

const negativeControls: readonly {
  readonly html: string;
  readonly id: string;
  readonly rule: Rule;
}[] = [
  { html: '<script>alert(1)</script>', id: 'forbidden-tag', rule: 'tag' },
  {
    html: '<p style="display:none">hidden</p>',
    id: 'smuggled-attribute',
    rule: 'attr',
  },
  {
    html: '<a href="javascript:alert(1)" rel="noopener noreferrer">click</a>',
    id: 'forbidden-href',
    rule: 'href',
  },
  {
    html: '<a href="https://example.com" rel="opener">click</a>',
    id: 'forged-rel',
    rule: 'rel',
  },
  {
    html: '<a href="https://example.com" rel="noopener noreferrer" target="_self">click</a>',
    id: 'forged-target',
    rule: 'target',
  },
  {
    html: '<svg><a href="https://example.com" rel="noopener noreferrer">click</a></svg>',
    id: 'foreign-namespace',
    rule: 'ns',
  },
];

interface AdversarialCase {
  readonly category: string;
  readonly expectedHref?: string | null;
  readonly expectedTarget?: '_blank' | null;
  readonly html: string;
  readonly id: string;
}

function joinHtml(...parts: readonly string[]): string {
  return parts.join('');
}

const adversarialCases: readonly AdversarialCase[] = [
  {
    category: 'mutation',
    html: joinHtml(
      '<noscript><p title="</noscript>',
      '<img src=x onerror=alert(1)>">safe</p></noscript>',
    ),
    id: 'noscript-parser-pivot',
  },
  {
    category: 'mutation',
    html: joinHtml(
      '<template><p><svg><style><img src=x onerror=alert(1)>',
      '</style></svg></p></template><p>safe</p>',
    ),
    id: 'template-foreign-content-pivot',
  },
  {
    category: 'mutation',
    html: joinHtml(
      '<math><mtext><table><mglyph><style><!--</style>',
      '<img title="--><img src=x onerror=alert(1)>">',
      '</mglyph></table></mtext></math>',
    ),
    id: 'math-mglyph-mutation-pivot',
  },
  {
    category: 'foreign-content',
    html: joinHtml(
      '<svg><foreignObject><math><mi>',
      '<a xlink:href="javascript:alert(1)">click</a>',
      '</mi></math></foreignObject></svg>',
    ),
    id: 'svg-math-xlink-pivot',
  },
  {
    category: 'scheme-obfuscation',
    expectedHref: null,
    html: '<a href="javascript%3Aalert(1)">click</a>',
    id: 'url-encoded-colon',
  },
  {
    category: 'scheme-obfuscation',
    expectedHref: null,
    html: '<a href="JaVa&#x53;CrIpT&colon;alert(1)">click</a>',
    id: 'mixed-entity-and-case',
  },
  {
    category: 'scheme-obfuscation',
    expectedHref: null,
    html: '<a href="java&#13;script&#58;alert(1)">click</a>',
    id: 'entity-carriage-return',
  },
  {
    category: 'attribute-smuggling',
    expectedHref: 'https://example.com/profile',
    expectedTarget: '_blank',
    html: joinHtml(
      '<a href="https://example.com/profile" target="_blank" rel="opener" ',
      'formaction="javascript:alert(1)" ',
      'srcdoc="<script>alert(1)</script>" ',
      'xlink:href="javascript:alert(1)" ',
      'style="background:url(javascript:alert(1))">click</a>',
    ),
    id: 'anchor-multi-attribute-smuggling',
  },
  {
    category: 'attribute-smuggling',
    html: joinHtml(
      '<p style="background:url(javascript:alert(1))" ',
      'onanimationstart="alert(1)" ',
      'srcdoc="<img src=x onerror=alert(1)>">safe</p>',
    ),
    id: 'plain-element-attribute-smuggling',
  },
  {
    category: 'namespace-confusion',
    html: joinHtml(
      '<math><a href="javascript:alert(1)" ',
      'rel="noopener noreferrer">click</a></math>',
    ),
    id: 'math-anchor-confusion',
  },
  {
    category: 'namespace-confusion',
    html: joinHtml(
      '<svg><a xlink:href="javascript:alert(1)" ',
      'target="_blank" rel="opener">click</a></svg>',
    ),
    id: 'svg-anchor-confusion',
  },
  {
    category: 'rel-target-forgery',
    expectedHref: 'https://example.com',
    expectedTarget: '_blank',
    html: joinHtml(
      '<a href="https://example.com" target="_blank" ',
      'rel="noopener opener noreferrer external">click</a>',
    ),
    id: 'extra-opener-rel-token',
  },
  {
    category: 'rel-target-forgery',
    expectedHref: 'https://example.com',
    expectedTarget: null,
    html: joinHtml(
      '<a href="https://example.com" target="_self" ',
      'rel="noopener noreferrer">click</a>',
    ),
    id: 'non-blank-target',
  },
  {
    category: 'rel-target-forgery',
    expectedHref: 'mailto:test@example.com',
    expectedTarget: null,
    html: joinHtml(
      '<a href="mailto:test@example.com" ',
      'target="_blank " rel="noreferrer">mail</a>',
    ),
    id: 'near-blank-target',
  },
];

type RandomSource = () => number;

function seededRandom(seed: number): RandomSource {
  let state = seed;

  return () => {
    state = (state * 1_664_525 + 1_013_904_223) % 0x1_0000_0000;
    return state / 0x1_0000_0000;
  };
}

function randomInteger(random: RandomSource, upperExclusive: number): number {
  return Math.floor(random() * upperExclusive);
}

const stringFragments = [
  '<',
  '>',
  '&',
  '"',
  String.fromCharCode(39),
  '=',
  '/',
  '\\',
  'javascript:',
  '&#x3a;',
  '</template>',
  '<svg/onload=alert(1)>',
  '\u0000',
  '\r\n\t',
] as const;

function arbitraryString(random: RandomSource): string {
  const fragmentCount = randomInteger(random, 96);
  let result = '';

  for (let index = 0; index < fragmentCount; index += 1) {
    if (randomInteger(random, 4) === 0) {
      result += stringFragments[randomInteger(random, stringFragments.length)];
    } else {
      result += String.fromCharCode(randomInteger(random, 0x1_0000));
    }
  }

  return result;
}

describe('independent rich-text neutralization predicate', () => {
  it.each(negativeControls)(
    'rejects live negative control $id',
    ({ html, rule }) => {
      const violations = neutralizationViolations(html);

      expect(isNeutralized(html)).toBe(false);
      expect(violations.map((violation) => violation.rule)).toContain(rule);
    },
  );

  it('rejects every structurally hostile raw corpus payload', () => {
    const structurallyHostile = HOSTILE_CORPUS.filter(
      ({ id }) => id !== 'js-scheme-bare',
    );

    expect(structurallyHostile).toHaveLength(HOSTILE_CORPUS.length - 1);
    for (const { id, payload } of structurallyHostile) {
      expect(neutralizationViolations(payload), id).not.toEqual([]);
    }
  });

  it('accepts dangerous-looking bare text as an inert text node', () => {
    const bareText = 'javascript:alert(1)';
    const document = parseRichText(bareText);

    expect(isNeutralized(bareText)).toBe(true);
    expect(document.body.childElementCount).toBe(0);
    expect(document.body.textContent).toBe(bareText);
  });

  it('accepts the complete safe tag and link surface', () => {
    const safe = joinHtml(
      '<p>text<br><strong>strong</strong><em>em</em><u>u</u></p>',
      '<ol><li>one</li></ol><ul><li>two</li></ul>',
      '<a href="tel:+84123456789" target="_blank" ',
      'rel="noopener noreferrer">call</a>',
    );

    expect(neutralizationViolations(safe)).toEqual([]);
  });
});

describe('client sanitizer adversarial cases', () => {
  it.each(HOSTILE_CORPUS)(
    'neutralizes corpus payload $id',
    ({ id, payload }) => {
      const output = sanitizeRichText(payload);

      expect(neutralizationViolations(output), id).toEqual([]);
      expect(sanitizeRichText(output), id).toBe(output);
    },
  );

  it.each(adversarialCases)(
    'neutralizes $category payload $id',
    ({ expectedHref, expectedTarget, html, id }) => {
      const output = sanitizeRichText(html);
      const anchors = anchorsIn(output);

      expect(neutralizationViolations(output), id).toEqual([]);
      expect(sanitizeRichText(output), id).toBe(output);
      for (const anchor of anchors) {
        expect(anchor.getAttribute('rel'), id).toBe(EXTERNAL_REL);
      }

      if (expectedHref !== undefined) {
        expect(anchors, id).toHaveLength(1);
        expect(anchors[0].getAttribute('href'), id).toBe(expectedHref);
      }
      if (expectedTarget !== undefined) {
        expect(anchors, id).toHaveLength(1);
        expect(anchors[0].getAttribute('target'), id).toBe(expectedTarget);
      }
    },
  );

  it('neutralizes seeded arbitrary strings and is idempotent', () => {
    const random = seededRandom(ARBITRARY_STRING_SEED);

    for (let index = 0; index < ARBITRARY_STRING_CASES; index += 1) {
      const input = arbitraryString(random);
      const output = sanitizeRichText(input);
      const caseLabel = `seed=${ARBITRARY_STRING_SEED} case=${index}`;

      expect(neutralizationViolations(output), caseLabel).toEqual([]);
      expect(sanitizeRichText(output), caseLabel).toBe(output);
    }
  });
});
